keel-ec2-plugin/src/main/kotlin/com/netflix/spinnaker/keel/ec2/resolvers/ImageResolver.kt

```diff
@@ -1,75 +1,58 @@
+import com.netflix.spinnaker.keel.api.ec2.ClusterSpec.LaunchConfigurationSpec
+import com.netflix.spinnaker.keel.api.ec2.ClusterSpec.ServerGroupSpec
+import com.netflix.spinnaker.keel.api.ec2.ClusterSpec.VirtualMachineImage
 import com.netflix.spinnaker.keel.api.ec2.JenkinsImageProvider
 import com.netflix.spinnaker.keel.api.id
-import com.netflix.spinnaker.keel.clouddriver.CloudDriverService
-import com.netflix.spinnaker.keel.clouddriver.DEFAULT_SERVICE_ACCOUNT
 import com.netflix.spinnaker.keel.clouddriver.ImageService
 import com.netflix.spinnaker.keel.clouddriver.model.NamedImage
 import com.netflix.spinnaker.keel.clouddriver.model.appVersion
-import com.netflix.spinnaker.keel.model.SubnetAwareRegionSpec
 import com.netflix.spinnaker.keel.persistence.ArtifactRepository
 import com.netflix.spinnaker.keel.persistence.DeliveryConfigRepository
+import com.netflix.spinnaker.keel.plugin.Resolver
 import com.netflix.spinnaker.kork.dynamicconfig.DynamicConfigService
+import kotlinx.coroutines.runBlocking
 import org.slf4j.LoggerFactory
+import org.springframework.stereotype.Component
 
+@Component
 class ImageResolver(
   private val dynamicConfigService: DynamicConfigService,
-  private val cloudDriverService: CloudDriverService,
   private val deliveryConfigRepository: DeliveryConfigRepository,
   private val artifactRepository: ArtifactRepository,
   private val imageService: ImageService
-) {
+) : Resolver<ClusterSpec> {
 
-  private val log by lazy { LoggerFactory.getLogger(javaClass) }
+  override val apiVersion: ApiVersion = SPINNAKER_API_V1.subApi("ec2")
+  override val supportedKind: String = "cluster"
 
-  // TODO: we may (probably will) want to make the parameter type more generic
-  suspend fun resolveImageId(resource: Resource<ClusterSpec>): ResolvedImages =
-    when (val imageProvider = resource.spec.imageProvider) {
-      is IdImageProvider -> resolveFromImageId(resource, imageProvider)
-      is ArtifactImageProvider -> resolveFromArtifact(resource, imageProvider)
-      is JenkinsImageProvider -> resolveFromJenkinsJob(resource, imageProvider)
-      else -> throw UnsupportedStrategy(
-        imageProvider::class.simpleName.orEmpty(),
-        ImageProvider::class.simpleName.orEmpty()
-      )
-    }
+  private val log by lazy { LoggerFactory.getLogger(javaClass) }
 
-  private suspend fun resolveFromImageId(
-    resource: Resource<ClusterSpec>,
-    imageProvider: IdImageProvider
-  ): ResolvedImages {
-    val images = cloudDriverService.namedImages(
-      serviceAccount = DEFAULT_SERVICE_ACCOUNT,
-      imageName = imageProvider.imageId,
-      account = null,
-      region = null
-    )
-    check(images.size == 1) {
-      "Expected a single image for AMI id ${imageProvider.imageId} but found ${images.size}"
+  override fun invoke(resource: Resource<ClusterSpec>): Resource<ClusterSpec> {
+    val imageProvider = resource.spec.imageProvider ?: return resource
+    val image = runBlocking {
```

Comment:
Curious about the use of `runBlocking` here... The docs say "it is designed to bridge regular blocking code to libraries that are written in suspending style, to be used in main functions and in tests". Are the `resolve*` function below written "in suspending style"?

Reply:
Never mind -- I need to do my homework and read more on this.

Reply:
yeah, the various calls to CloudDriver are suspending functions.